File: src/0806_Hanota_LCCI.py
# 0806_Hanota_LCCI
from typing import List

class Solution:
    # Method to initiate the Tower of Hanoi process.
    def hanota(self, A: List[int], B: List[int], C: List[int]) -> None:
        n = len(A)  # Get the number of disks in the initial tower (A).
        self.move(n, A, B, C)  # Begin the process of moving disks.

    # Recursive method to move disks between towers.
    def move(self, n, A, B, C):
        if n == 1:
            # Base case: Directly move a single disk from A to C.
            C.append(A[-1])  # Append the top element of A to C.
            A.pop()  # Remove the top element from A.
            return
        else:
            # Recursive case: Move n-1 disks from A to B, using C as auxiliary.
            self.move(n-1, A, C, B)

            # Move the remaining disk from A to C.
            C.append(A[-1])  # Append the top element of A to C.
            A.pop()  # Remove the top element from A.

            # Move the n-1 disks from B to C, using A as auxiliary.
            self.move(n-1, B, A, C)

File: src/0806_Hanota_LCCI.c
// Define a structure for representing a frame in the custom stack.
typedef struct
{
    int pc;             // Program counter to track the next operation
    int n;              // Number of disks
    char from, to, via; // Identifiers for the rods (source, destination, auxiliary)
} Frame;

// Macro to simulate a function call, pushing a new frame onto the stack.
#define call(...) ({ *(++top) = (Frame){.pc = 0, __VA_ARGS__}; })

// Macro to simulate a function return, popping a frame from the stack.
#define ret() ({ top--; })

// Macro to simulate a goto operation, modifying the program counter of the current frame.
#define goto(loc) ({ f->pc = (loc)-1; })

// Function to solve the Tower of Hanoi problem iteratively.
void hanoi(int n, char from, char to, char via)
{
    Frame stk[64], *top = stk - 1; // Stack initialization with a fixed size and top pointer

    // Simulate the initial function call
    call(n, from, to, via);

    // Main loop for processing each frame on the stack
    for (Frame *f; (f = top) >= stk; f->pc++)
    {
        // Load the frame's context into local variables
        n = f->n;
        from = f->from;
        to = f->to;
        via = f->via;

        // Switch-case to determine the next action based on the program counter
        switch (f->pc)
        {
        case 0:
            if (n == 1) // Base case: move a single disk directly
            {
                printf("%c -> %c\n", from, to);
                goto(4); // Skip to the end of the function
            }
            break;
        case 1:
            call(n - 1, from, via, to); // Simulate recursive call to move n-1 disks
            break;
        case 2:
            call(1, from, to, via); // Move the remaining disk
            break;
        case 3:
            call(n - 1, via, to, from); // Move the n-1 disks to the target rod
            break;
        case 4:
            ret(); // Return from the function call
            break;
        default:
            assert(0); // Assert failure in case of an unexpected program counter value
        }
    }
}
